File: 3150-275-916-word-subsets/3150-275-916-word-subsets.cpp
class Solution {
public:
    vector<int>solve(string &s){
        vector<int>v(26);
        for(char ch : s)
            v[ch-'a']++;
        return v;
    }
    vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
        vector<int>frq(26);
        for(string s : words2){
            vector<int>tmp = solve(s);
            for(int i = 0 ; i <26 ;i++)
                frq[i] = max(frq[i],tmp[i]);
        }
        vector<string>ans;
        for(string s : words1){
            vector<int>tmp = solve(s);
            bool ok = true;
            for(int i = 0 ; i < 26 ;i++)
                if(frq[i]> tmp[i])
                    ok = false;

            if(ok)ans.push_back(s);
        }
        return ans;


    }
};

File: 3150-275-916-word-subsets/3150-275-916-word-subsets.py
class Solution:
    def solve(self, s: str) -> list[int]:
        freq = [0] * 26
        for ch in s:
            freq[ord(ch) - ord('a')] += 1
        return freq

    def wordSubsets(self, words1: list[str], words2: list[str]) -> list[str]:
        frq = [0] * 26
        for word in words2:
            tmp = self.solve(word)
            for i in range(26):
                frq[i] = max(frq[i], tmp[i])

        result = []
        for word in words1:
            tmp = self.solve(word)
            if all(frq[i] <= tmp[i] for i in range(26)):
                result.append(word)

        return result
